namecache/namecache.go

```diff
@@ -0,0 +1,151 @@
+// Package namecache implements background following (resolution and pinning) of names
+package namecache
+
+import (
+	"context"
+	"strings"
+	"sync"
+	"time"
+
+	merkledag "github.com/ipfs/go-ipfs/merkledag"
+	namesys "github.com/ipfs/go-ipfs/namesys"
+	path "github.com/ipfs/go-ipfs/path"
+	pin "github.com/ipfs/go-ipfs/pin"
+	uio "github.com/ipfs/go-ipfs/unixfs/io"
+
+	logging "gx/ipfs/QmSpJByNKFX1sCsHBEp3R73FL4NF6FnQTEGyNAXHm2GS52/go-log"
+)
+
+const (
+	followInterval = 60 * time.Minute
+)
+
+var log = logging.Logger("namecache")
+
+// NameCache represents a following cache of names
+type NameCache interface {
+	Follow(name string, pinit bool)
+	Unfollow(name string)
+	ListFollows() []string
+}
+
+type nameCache struct {
+	nsys    namesys.NameSystem
+	pinning pin.Pinner
+	dag     merkledag.DAGService
+
+	ctx     context.Context
+	follows map[string]func()
+	mx      sync.Mutex
+}
+
+func NewNameCache(ctx context.Context, nsys namesys.NameSystem, pinning pin.Pinner, dag merkledag.DAGService) NameCache {
+	return &nameCache{
+		ctx:     ctx,
+		nsys:    nsys,
+		pinning: pinning,
+		dag:     dag,
+		follows: make(map[string]func()),
+	}
+}
+
+// Follow spawns a goroutine that periodically resolves a name
+// and (when pinit is true) pins it in the background
+func (nc *nameCache) Follow(name string, pinit bool) {
+	nc.mx.Lock()
+	defer nc.mx.Unlock()
+
+	if _, ok := nc.follows[name]; ok {
+		return
+	}
+
+	ctx, cancel := context.WithCancel(nc.ctx)
+	go nc.followName(ctx, name, pinit)
+	nc.follows[name] = cancel
+}
+
+// Unfollow cancels a follow
+func (nc *nameCache) Unfollow(name string) {
+	nc.mx.Lock()
+	defer nc.mx.Unlock()
+
+	cancel, ok := nc.follows[name]
+	if ok {
+		cancel()
+		delete(nc.follows, name)
+	}
+}
+
+// ListFollows returns a list of names currently being followed
+func (nc *nameCache) ListFollows() []string {
+	nc.mx.Lock()
+	defer nc.mx.Unlock()
+
+	follows := make([]string, 0)
+	for name, _ := range nc.follows {
+		follows = append(follows, name)
+	}
+
+	return follows
+}
+
+func (nc *nameCache) followName(ctx context.Context, name string, pinit bool) {
```

Comment:
Having this as a single worker resolving one name each `followInterval/len(nc.follows)` will ensure no activity spikes when many names are followed.

Reply:
I can coalesce in a single goroutine, more resource economical.

Reply:
I am not sure there is significant benefit by having a single goroutine, plus it's a lot more complicated to write if we want to ensure immediate initial resolution upon follow.
The pins will be serialized by the pin lock.

How strongly do you feel about using a single goroutine?

Reply:
It was more of a suggestion, though it's still something we may want to do in the feature